map.go

```diff
@@ -1,71 +1,71 @@
 package hamt
 
 // Map represents a map (associative array).
-type Map struct {
-	set Set
+type Map[K Entry[K], V any] struct {
+	set Set[keyValue[K, V]]
 }
 
 // NewMap creates a new map.
-func NewMap() Map {
-	return Map{NewSet()}
+func NewMap[K Entry[K], V any]() Map[K, V] {
+	return Map[K, V]{NewSet[keyValue[K, V]]()}
 }
 
 // Insert inserts a key-value pair into a map.
-func (m Map) Insert(k Entry, v interface{}) Map {
-	return Map{m.set.Insert(newKeyValue(k, v))}
+func (m Map[K, V]) Insert(k K, v V) Map[K, V] {
+	return Map[K, V]{m.set.Insert(newKeyValue(k, v))}
 }
 
 // Delete deletes a pair of a key and a value from a map.
-func (m Map) Delete(k Entry) Map {
-	return Map{m.set.Delete(k)}
+func (m Map[K, V]) Delete(k K) Map[K, V] {
+	var zero V
+	return Map[K, V]{m.set.Delete(newKeyValue(k, zero))}
 }
 
 // Find finds a value corresponding to a given key from a map.
 // It returns nil if no value is found.
-func (m Map) Find(k Entry) interface{} {
-	e := m.set.find(k)
+func (m Map[K, V]) Find(k K) *V {
+	var zero V
+	e := m.set.find(newKeyValue(k, zero))
 
 	if e == nil {
 		return nil
 	}
 
-	return e.(keyValue).value
+	return &e.value
 }
 
 // Include returns true if a key-value pair corresponding with a given key is
 // included in a map, or false otherwise.
-func (m Map) Include(k Entry) bool {
+func (m Map[K, V]) Include(k K) bool {
 	return m.Find(k) != nil
 }
 
 // FirstRest returns a key-value pair in a map and a rest of the map.
 // This method is useful for iteration.
 // The key and value would be nil if the map is empty.
-func (m Map) FirstRest() (Entry, interface{}, Map) {
+func (m Map[K, V]) FirstRest() (*K, *V, Map[K, V]) {
```

Comment:
I was thinking of this and `Find` operations for map types. Should we actually delegate the boxing of key and value types to users? This actually allows in-place modification of keys although it's invalid and I'm not sure if it's an intuitive interface. WDYT?

Reply:
Ah Go doesn't have `Option`. My brain is so Rusty now... orz

Reply:
We could add an extra `ok` return I suppose, by analogy to how built-in maps do this?

Reply:
Yeah. That sounds like a good alternative although I probably made it return values directly so that users can write codes in a more functional way (e.g. `f(map.Find(k))`) without checking flags every time.